pytorch_pfn_extras/training/extensions/slack.py

```diff
@@ -0,0 +1,300 @@
+import getpass
+import os
+import json
+import urllib.request
+import shlex
+import sys
+import socket
+import warnings
+
+from typing import Any, Callable, Dict, List, Optional, Union, TYPE_CHECKING
+
+from pytorch_pfn_extras.training import extension
+from pytorch_pfn_extras.training._manager_protocol import ExtensionsManagerProtocol
+
+
+if TYPE_CHECKING:
+    from pytorch_pfn_extras.training._trigger_util import TriggerLike
+
+
+try:
+    import slack_sdk
+    _slack_sdk_available = True
+except ImportError:
+    _slack_sdk_available = False
+
+
+_identity = f'{getpass.getuser()}@{socket.gethostname()} [PID {os.getpid()}]'
+
+
+def _default_start_msg(m: ExtensionsManagerProtocol, c: Any, o: Dict[Any, Any]) -> str:
+    return f'''**Training started! {_identity}**
+Command: `{shlex.quote(' '.join(sys.argv))}`
+'''
+
+
+_default_end_msg = f'**Training finished! {_identity}**'
+
+
+class Slack(extension.Extension):
+    """An extension to communicate with Slack.
+
+    This extension receives a ``msg`` argument that contains
+    placeholders that will be populated with ``manager`` or ``manager.observation``
+    fields, and custom objects such as ``context_object``.
+
+    For example. `msg = "Loss {loss} for iteration {.iteration}"`
+    will be populated as `msg.format(manager, context_object, **observation)`
+    retrieving the ``loss`` value from the ``observation`` and ``.iteration`` from
+    the manager object. Instead of string, a callable object taking the
+    ``ExtensionsManager``, ``context_object`` and the observation dictionary can
+    be used instead. The same rules apply for the ``filenames_template`` argument.
+
+    Args:
+        channel_id (str): The channel where messages or files will be sent.
+        msg (str or callable): Template for sending the message.
+            It can be a string to be formatted using ``.format`` or a callable
+            that returns a string. In both cases, `manager`, the current
+            observation dict and the context object are used to look up
+            values of interest.
+        start_msg (str or callable): Template for sending a message
+            at the beggining of the experiment. If ``None``, the default
+            start message will be sent. To avoid sending a message use
+            ``''``.
+            See ``msg`` for format.
+        end_msg (str or callable): Template for sending a message
+            at the completion of the experiment. If ``None``, the default
+            end message will be sent. To avoid sending a message use
+            See ``msg`` for format.
+        filenames(list of str or callable): list of files that will
```

Comment:
```suggestion
        filenames (list of str or callable): list of files that will
```

Reply:
* I think users will likely want a trigger to decide when to upload files. (I guess in most cases it is set to the max_epochs)
* `start_msg` should not upload files.